src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -583,22 +583,28 @@ private PollingResult compareRemoteRevisionWithImpl(Job<?, ?> project, Launcher
                     for (BranchSpec branchSpec : getBranches()) {
                         for (Entry<String, ObjectId> entry : heads.entrySet()) {
                             final String head = entry.getKey();
-                            String name;
                             // head is "refs/(heads|tags)/branchName
-                            if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
-                            else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
-                            else name = remote + "/" + head;
 
-                            if (!branchSpec.matches(name, environment)) continue;
+                            // first, check the a canonical git reference is configured
+                            if (!branchSpec.matches(head, environment)) {
+
+                                // convert head `refs/*` into shortcut notation `remote/branch`
+                                String name = head;
+                                if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
+                                else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
```

Comment:
@ndeloof you can even do github PR builds without any PR plugin, github stores all PRs in refs/pull/$ID/{head,merge} . This is not documented but this is how all gh plugins works https://github.com/jenkinsci/github-pullrequest-plugin/blob/master/README.md (Configuration part).

Also https://github.com/jenkinsci/github-plugin/pull/39 probably unrelated, but in general refs hardcodes raising many questions.